### PR TITLE
Build PyPI package

### PR DESCRIPTION
**Pull request by mary-cleaton**
## Description

Update GitHub Actions workflows so that on push to `main` a package will be built and published on PyPI. This is still somewhat hypothetical as until we test it we don't know if it works!

Also updated a lot of documentation to reflect this change to the repository's deployment.

## Type of change

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [x] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: .github/workflows/pull_request.yaml [hunk 1]
```diff
@@ -8,51 +8,63 @@
...
       
       - name: Run pre-commit hooks
-        uses: pre-commit/action@v3.0.1
+        id: pre-commit-run
+        uses: pre-commit/action@2c7b3805fd2a0fd8c1884dcaebf91fc102a13ecd
 
-  # Enforces the running of all unit tests on every pull request
   test:
-    needs: changelog
+    name: Run unit tests
```
[the remark this answers is not in the document]
Comment (mary-cleaton) [reply .github/workflows/pull_request.yaml:52]: Fixed this.

File: .github/workflows/increment_version.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      - name: Check version
+        id: version-check
+        uses: maybe-hello-world/pyproject-check-version@cb964e82b3c8a0b1cedb117da63e1761c46600b8
+        with:
+          pyproject-path: "./pyproject.toml"
+
+      - name: Set version as env variable
+        id: version-set
+        run: |
+          echo "PACKAGE_VERSION=${{ steps.versioncheck.outputs.local_version }}" >> $GITHUB_ENV
```
Comment (danasebo) [inline .github/workflows/increment_version.yaml]: shouldn't this be `steps.version-check.outputs[...]` as it relates to step with id `version-check`? or it doesn't matter?

File: .github/workflows/deploy_to_test_pypi.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      - name: Retrieve release distributions
+        id: publish-retrieve-dist
+        uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c
+        with:
+          name: release-dists
+          path: dist/
+
+      - name: Publish release distributions on PyPI
+        id: publish-to-pypi
+        uses: pypa/gh-action-pypi-publish@cef221092ed1bacb1cc03d23a2d87d1d172e277b
```
Comment (danasebo) [inline .github/workflows/deploy_to_test_pypi.yaml:50]: are you sure this will publish to test pypi? I checked this action and it looks like it will publish to pypi by default, and some additional settings might be needed if it wants to be published to test pypi (see https://github.com/pypa/gh-action-pypi-publish#trusted-publishing). I know you specify test-pypi environment above but I am not sure if that's sufficient.

File: .github/workflows/deploy_to_pypi.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+jobs:
+  release-build:
+    name: Build and verify release distributions
+    runs-on: ubuntu-latest
+    permissions:
+      # allows the provenance attestation to work
+      attestations: write
+      id-token: write
+    # only allow tagged commits to be deployed
```
Comment (danasebo) [inline .github/workflows/deploy_to_pypi.yaml]: what happens if this is not the case? don't you want to include a clear failure message step so the user knows what happened?

File: .github/workflows/deploy_to_pypi.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
+name: Deploy version to PyPI
+
+on: 
+  workflow_dispatch:
+    inputs:
+      tags:
+        description: "Version to deploy, e.g. 1.0.0"
+        required: true
+        type: string
```
Comment (danasebo) [inline .github/workflows/deploy_to_pypi.yaml:9]: I don't really get how / when this is used, sorry! Can you please explain?

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -1,55 +1,77 @@
...
 
   - What is this change?
   - Is this a bug fix or a feature and does it break any existing functionality?
   - How has it been tested?
+
 </details>
  
-*This pr introduces....*
-## Type of change
+**This pr introduces....**
```
Comment (danasebo) [inline .github/pull_request_template.md]: Suggestion (feel free to ignore) `This pr [...]` -> `This PR [...]`

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -1,55 +1,77 @@
...
-- [ ] Bug fix - *non-breaking change*
-- [ ] New feature - *non-breaking change*
-- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
-- [ ] Non-user facing change, structural change, dev functionality, docs ...
+- [ ] Bug fix.
+- [ ] New feature.
+- [ ] Breaking change - **backwards incompatible change**, changes expected behaviour.
+- [ ] Non-user facing change - structural change, dev functionality, docs, etc.
 
 ## Checklist:
```
Comment (danasebo) [inline .github/pull_request_template.md]: Suggestion (feel free to ignore) `Checklist:` -> `Checklist` (to be consistent with other sections)

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -1,55 +1,77 @@
...
-- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
-- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
-- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.
+You should review all of the following:
+
+- **Documentation**:
+  - Are docstrings present?
+  - Are comments only present where necessary and have they been added/updated?
+- **Style guidelines**:
+  - Do the new files conforms to the project's contribution guidelines?
```
Comment (danasebo) [inline .github/pull_request_template.md]: conforms -> conform

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -1,55 +1,77 @@
...
+- **Functionality**:
+  - Does the code works as expected?
```
Comment (danasebo) [inline .github/pull_request_template.md]: works -> work

File: .github/pull_request_template.md [hunk 10]
```diff
@@ -1,55 +1,77 @@
...
+  - Does the code handle expected edge cases and exceptions appropriately?
+- **Complexity**:
+  - Is the code is not overly complex?
+  - Has the logic been split into appropriately sized functions?
+- **Test coverage**:
+  - Are there unit tests for all essential functions for a reasonable range of inputs and conditions?
+  - Do added and existing tests locally on the reviewer's machine?
```
Comment (danasebo) [inline .github/pull_request_template.md]: existing tests locally -> existing tests pass locally

File: .github/pull_request_template.md [hunk 11]
```diff
@@ -1,55 +1,77 @@
...
+- [ ] Performed a self-review of my own code.
+- [ ] Commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
+- [ ] Made corresponding changes to the documentation (comments, docstring, etc.).
+- [ ] Added tests that prove my fix is effective or that my feature works.
+- [ ] Checked that new and existing unit tests pass locally with my changes.
+- [ ] Updated the change log.
 
-#  Peer review
-Any new code includes all the following:
+##  Peer review
```
Comment (danasebo) [inline .github/pull_request_template.md]: remove extra space before `Peer`

File: .github/pull_request_template.md [hunk 12]
```diff
@@ -1,55 +1,77 @@
...
+- **Documentation**:
+  - Are docstrings present?
+  - Are comments only present where necessary and have they been added/updated?
+- **Style guidelines**:
+  - Do the new files conforms to the project's contribution guidelines?
+- **Functionality**:
+  - Does the code works as expected?
+  - Does the code handle expected edge cases and exceptions appropriately?
+- **Complexity**:
+  - Is the code is not overly complex?
```
Comment (danasebo) [inline .github/pull_request_template.md]: is not -> not

File: scalelink/__init__.py [hunk 13]
```diff
@@ -1 +1 @@
-__version__ = "0.1.1"
+from .main import run_scalelink as run_scalelink
```
Comment (danasebo) [inline scalelink/__init__.py]: why `import run_scalelink as run_scalelink` is required, and not simple `import run_scalelink`?

File: scalelink/main.py
Comment (danasebo) [inline scalelink/main.py]: please update the return type hint as the function now does not return anything.

File: pyproject.toml [hunk 14]
```diff
@@ -1,3 +1,76 @@
...
+version = "0.1.1"
+dependencies = [
+  "boto3 >= 1.34.149",
+  "numpy >= 1.24",
+  "pandas >= 2.0",
+  "pyspark >= 3.0.0, < 4.0.0",
+  "raz_client >= 1.1.0",
+  "rdsa_utils >= 0.16.0",
+]
+requires-python = ">= 3.10"
```
Comment (danasebo) [inline pyproject.toml]: previous implementation (using setup.cfg file) required Python >=3.10, <3.12. Are you sure you don't want to use <3.12 requirement?

**Review by danasebo — CHANGES_REQUESTED** [review]
I am not able to test whether the github workflows work as intended, but I have reviewed them to the best of my knowledge. There are a few comments to address to improve readability and understanding, otherwise the changes look really good to me- excellent work, Mary!

I'll submit additional comments on the user guides and a couple of the remaining files shortly.

File: .github/workflows/increment_version.yaml (hunk 2 above)
Comment (mary-cleaton) [reply .github/workflows/increment_version.yaml]: You are completely correct! Fixed now.

File: .github/workflows/deploy_to_test_pypi.yaml (hunk 3 above)
Comment (mary-cleaton) [reply .github/workflows/deploy_to_test_pypi.yaml:50]: I wasn't sure if that was needed or if the `url` in the `environment` setting was sufficient. Probably best to go for belt and braces, though. I've changed it as per your suggestion.

File: .github/workflows/deploy_to_pypi.yaml (hunk 4 above)
Comment (mary-cleaton) [reply .github/workflows/deploy_to_pypi.yaml]: GitHub Actions `if` doesn't have an else condition. So, I can't make a custom failure message without using shell script (which a) I'm bad at and b) would make things more complex than I think they need to be). If it fails, I'm pretty sure it will report that the GitHub action fails due to that line. Between that automatic fail message and the comment in the workflow script I think it's reasonably comprehensible (at least for a first attempt at a deployment workflow).

File: .github/workflows/deploy_to_pypi.yaml (hunk 5 above)
Comment (mary-cleaton) [reply .github/workflows/deploy_to_pypi.yaml:9]: Ok, so the `workflow_dispatch` spec means the [workflow is manually deployed](https://docs.github.com/en/actions/how-tos/manage-workflow-runs/manually-run-a-workflow). As per that link, once the workflow is in the default branch you can access it in the Actions tab of the repo and see a "Run workflow" button.

What the rest of the spec does (if it works!) is provide a text box so that you don't just deploy the `main` branch but you deploy a specific version from the `main` branch. The text box is titled with the `description`, must be filled out (due to the `required` var being `true`) and is a string. Alternatively, you can put something else in the `type` option and get a drop-down menu or a text box that only accepts integers or whatever.

I think this should work, but I'm kinda going from memory for how the one someone did for me in DSIT works, so I may have not quite got the syntax right.

File: .github/pull_request_template.md (hunk 6 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed.

File: .github/pull_request_template.md (hunk 7 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed.

File: .github/pull_request_template.md (hunk 8 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed.

File: .github/pull_request_template.md (hunk 9 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed.

File: .github/pull_request_template.md (hunk 10 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed.

File: .github/pull_request_template.md (hunk 11 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed.

File: .github/pull_request_template.md (hunk 12 above)
Comment (mary-cleaton) [reply .github/pull_request_template.md]: Fixed. Sorry about all these. Was modifying the template from RDSA to something more applicable to our needs but evidently didn't do a great job of modifying!

File: scalelink/__init__.py (hunk 13 above)
Comment (mary-cleaton) [reply scalelink/__init__.py]: Ah, cause I got myself muddled. It should have been either `import .main.run_scalelink as run_scalelink` or `from .main import run_scalelink`. Fixed.

File: scalelink/main.py
Comment (mary-cleaton) [reply scalelink/main.py]: Done.

File: pyproject.toml (hunk 14 above)
Comment (mary-cleaton) [reply pyproject.toml]: Good point. We've only tested it on 3.10 and 3.11 so it should be <3.12.
Fixed.

File: .github/workflows/deploy_to_pypi.yaml (hunk 4 above)
Comment (danasebo) [reply .github/workflows/deploy_to_pypi.yaml]: In your `.github/workflows/pull_request.yaml` script (line 22) you use `echo "ERROR: You can only merge to main from develop."` - isn't that what we need here as well?

Comment (danasebo) [reply .github/workflows/deploy_to_pypi.yaml]: but if it is more complicated than this, happy for you to go ahead as it is.

File: scalelink/main.py
Comment (danasebo) [inline scalelink/main.py]: please remove

File: pyproject.toml [hunk 15]
```diff
@@ -1,3 +1,76 @@
...
+version = "0.1.1"
+dependencies = [
+  "boto3 >= 1.34.149",
+  "numpy >= 1.24",
+  "pandas >= 2.0",
+  "pyspark >= 3.0.0, < 4.0.0",
+  "raz_client >= 1.1.0",
+  "rdsa_utils >= 0.16.0",
+]
+requires-python = ">= 3.10,< 3.12"
```
Comment (danasebo) [inline pyproject.toml]: silly thing but I noticed it and it will take you 2secs to implement, so here we go `,<` -> `, <` (i.e. please add space after comma)

File: pyproject.toml [hunk 16]
```diff
@@ -1,3 +1,76 @@
 [build-system]
```
Comment (danasebo) [inline pyproject.toml:1]: Can you please confirm that you tested the building using `pyproject.toml` and that it works? I am unable to test it as I don't have a CDP dev access at the moment.

File: CHANGELOG.md [hunk 17]
```diff
@@ -44,8 +43,15 @@ and this project adheres to [semantic versioning](https://semver.org/spec/v2.0.0
...
-  - Added contact, dedication and icons for linter and formatter.
+  - Added installation and use, contact, dedication and icons.
   - Edited pre-requisites and acknowledgements.
+- Branch and Deploy Guide:
+  - Updated description of GitHub Actions.
+  - Updated deployment instructions.
+- Build method, from using `setuptools` with `setup.py`, `setup.cfg` and `requirements.txt` to using `hatchling` with `pyproject.toml`.
+- Location of config template, so it is included in build.
+- Minor phrasing changes to User Guide (`docs/user_guide.md`) and pull request template (`.github/pull_request_template.md`).
+- Changed all third-party resuable workflows in GitHub Actions workflows to pin to latest release's long commit hash.
```
Comment (danasebo) [inline CHANGELOG.md]: resuable -> reusable

File: CHANGELOG.md [hunk 18]
```diff
@@ -60,6 +66,8 @@ and this project adheres to [semantic versioning](https://semver.org/spec/v2.0.0
 ### Removed
 
 - Python 3.8 and 3.9 support.
+- Use of `bump2version`, as it's no longer supported.
+- Use of `hotfix` branches by this project. Updated Branch and Deploy Guide and Pull Request GitHub Action workflow to reflec this.
```
Comment (danasebo) [inline CHANGELOG.md]: reflec -> reflect

File: CHANGELOG.md [hunk 19]
```diff
@@ -44,8 +43,15 @@ and this project adheres to [semantic versioning](https://semver.org/spec/v2.0.0
   - `test_get_input_variables` in `tests/utils/utils.py`.
```
Comment (danasebo) [inline CHANGELOG.md]: in `tests/utils/utils.py`. -> in `tests/utils/test_utils.py`.

File: CHANGELOG.md
Comment (danasebo) [inline CHANGELOG.md]: in tests/utils/utils.py. -> in tests/utils/test_utils.py.

Comment (danasebo) [inline CHANGELOG.md]: in tests/utils/utils.py. -> in tests/utils/test_utils.py.

File: docs/branch_and_deploy_guide.md [hunk 20]
```diff
@@ -65,150 +62,278 @@ Our repository has two permanent branches:
...
+   - When a `push` to `main` is detected.
 
-2. **Increment version:**
-   - The version of the package is incremented.
-   - The new version tag is pushed.
-   - A new GitHub release is made that refers to the new version tag.
-  
-Further deployment steps will be added in due course.
+2. **Extract repo version:**
+   - Extract the version of the repo from the location specified in `setup.py`.
```
Comment (danasebo) [inline docs/branch_and_deploy_guide.md]: setup.py is no longer user, please update

File: docs/branch_and_deploy_guide.md [hunk 21]
```diff
@@ -65,150 +62,278 @@ Our repository has two permanent branches:
...
+2. **Upload the test build to Test PyPI:**
+   - Upload by running: `twine upload -r testpypi dist/*`.
+   - When prompted, input your Test PyPI API token.
+   
+3. **Check the package styling:**
+   - Check the uploaded package on Test PyPI by following the link provided in the terminal.
+   - Review the styling of the information from `README.md`. Make a note of any changes that need to be implemented
+     prior to uploading to PyPI.
+
+3. **Download from Test PyPI and test:**
```
Comment (danasebo) [inline docs/branch_and_deploy_guide.md]: I know it does not matter in .md, but for consistency, this should be 4

File: docs/branch_and_deploy_guide.md [hunk 22]
```diff
@@ -65,150 +62,278 @@ Our repository has two permanent branches:
...
+   - Manually update the package version in `pyproject.toml` following [semantic versioning principles][sem-ver].
+
+2. **Update `CHANGELOG.md`:**
+   - Continue to work in the `develop` branch.
+   - In the `CHANGELOG.md` file, create a new header/section for the newly bumped version.
+   - Move all entries from the "Unreleased" section to the new version section. This action effectively transfers the
+     documentation of changes from being pending release to being part of the new version's official changelog.
+   - Ensure the "Unreleased" section is left empty after this process, ready for documenting future changes.
+   - Update the "Release links" section at the bottom of the document. Add links to the new version's GitHub Release
+     page and its PyPi listing, following the existing format. **Note: this repo does not currently have a PyPi listing.**
```
Comment (danasebo) [inline docs/branch_and_deploy_guide.md]: PyPi -> PyPI (twice in this bullet point)

File: docs/branch_and_deploy_guide.md [hunk 23]
```diff
@@ -65,150 +62,278 @@ Our repository has two permanent branches:
...
+    Feat4[Raise pull request to merge feature branch into develop branch]
+    Feat5[Trigger automated checks via GitHub Actions]
+    Feat6[Review pull request]
+    Feat7{Feature branch: approve pull request?}
+    Feat8[Merge pull request]
+    
+    Dev1{Develop branch: Ready for release?}
+    Dev2[Manually create build dist locally]
+    Dev3[Deploy build to Test PyPI]
+    Dev4[Check description and metadata is correct on Test PyPI]
```
Comment (danasebo) [inline docs/branch_and_deploy_guide.md]: is -> are

File: docs/user_guide.md [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+| filepaths | ssl_file           | Y         | The relative filepath to your local SSL Certificate file, ending `.crt`. For `raz_client`, to permit use of `boto3`.                                                                                                                                                                                                                   |
+| filepaths | df1_path           | N*        | The relative filepath to the first dataset you wish to link. This must be a parquet file in your AWS S3 bucket.                                                                                                                                                                                                                        |
+| filepaths | df2_path           | N*        | The relative filepath to the second dataset you with to link. This must be a parquet file in your AWS S3 bucket.                                                                                                                                                                                                                       |
```
Comment (danasebo) [inline docs/user_guide.md]: you with to link -> you wish to link

File: docs/user_guide.md [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+| cutpoints | \<see contents\>   | Y         | There must be the same number of configs in this section as there are linkage variables. Each must be named `<linkage_var>_cutpoints`. If the linkage variable has cutpoints, these must be entered smallest to largest as a comma-separated string. If the linkage variable does not have cutpoints, the config value must be `None`. |
+
+\* You can either use `df1_path` and `df2_path` to supply the two datasets to be linked or you can supply these datasets already joined using `df_candidates_path`. If you
+do the former, `run_scalelink` will Cartesian-join your datasets before calculating Scalelink match scores. If you do the latter, you can provide an alternative, e.g. a
+blocked set of candidate pairs. However, note that it is currently unknown how blocking interacts with the Scalelink method.
+
+### Configs example
+
+Imagine you had an AWS S3 bucket named `s3a://my-aws-s3-bucket`. Inside this bucket are two parquet files at `/my_project/dataframe1` and `/my_project/dataframe2`.
+You decide to let the `run_scalelink`
```
Comment (danasebo) [inline docs/user_guide.md]: this looks like an incomplete sentence

File: docs/user_guide.md [hunk 26]
```diff
@@ -0,0 +1,142 @@
+# User guide
+
+## Pre-requisites
+
+See [README][readme].
```
Comment (danasebo) [inline docs/user_guide.md:5]: this link does not work, I guess README is somewhere else now?

File: docs/user_guide.md [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+
+run_scalelink(config_path = "<absolute/path/to/config/file.ini")
+```
+
+This requires an appropriately set-up config file.
+
+### Configs
+
+The config file for `scalelink` is a `.ini` file. You can name it whatever you choose and store it wherever you choose, provided that you reference that file name and
+path in the `config_path` argument when using the `run_scalelink` function. A [config file template][config-template] is provided to guide you in creating your config
```
Comment (danasebo) [inline docs/user_guide.md:28]: also the link for the config template does not work after the config template was moved to a different location

**Review by danasebo — CHANGES_REQUESTED** [review]

File: docs/user_guide.md [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+- You have a local SSL Certificate and you know the relative filepath to it.
+- Your input datasets are saved as parquet files in a location within your AWS S3 bucket.
+- Your linkage variables are numeric or string. Linkage variables that are any other type (including date, timestamp or boolean) are not currently permitted.
+  - All numeric linkage variables are compared on a binary basis - i.e. are they exactly the same or not?
+  - All string linkage variables are compared using the Sorensen-Dice coefficient only. Other string comparison methods are not currently available.
+- Your categorical linkage variables (e.g. `sex`) have been label-encoded to turn them into numeric variables.
+- Your linkage variables do not contain any missingness (either as `None` or as a place-holder, e.g. a string consisting of `"None"`, `"N/A"`, etc.).
+- You want your linked dataset to be outputted as a parquet file saved to a location within your AWS S3 bucket.
+
+[config-template]: scalelink/configs_template.ini
```
Comment (danasebo) [inline docs/user_guide.md]: I guess it should be a relative path to this user guide, i.e. ../scalelink/configs_template.ini in order for the reference to work?

File: docs/user_guide.md [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+[github-releases]: https://github.com/ONSdigital/scalelink/releases
+[pypi]: https://pypi.org/project/scalelink/
+[readme]: README.md
```
Comment (danasebo) [inline docs/user_guide.md]: same here, probably it should be ../README.md

File: scalelink/main.py
Comment (mary-cleaton) [reply scalelink/main.py]: Done.

File: pyproject.toml (hunk 15 above)
Comment (mary-cleaton) [reply pyproject.toml]: Done.

File: pyproject.toml (hunk 16 above)
Comment (mary-cleaton) [reply pyproject.toml:1]: Yes, that's how I did the current Test PyPI deployment. The only bit that's untested is using the GitHub Actions workflow to deploy (to Test PyPI or PyPI) rather than using command line (as was used for the current Test PyPI deployment).

File: CHANGELOG.md (hunk 17 above)
Comment (mary-cleaton) [reply CHANGELOG.md]: Thanks! Fixed.

Clearly I need to put a spellchecker in the pre-commit hooks! Did that in DSIT and it worked really well and got rid of all this sort of nonsense.

File: CHANGELOG.md (hunk 18 above)
Comment (mary-cleaton) [reply CHANGELOG.md]: Done.

File: CHANGELOG.md (hunk 19 above)
Comment (mary-cleaton) [reply CHANGELOG.md]: Fixed.

File: CHANGELOG.md
Comment (mary-cleaton) [reply CHANGELOG.md]: Fixed.

Comment (mary-cleaton) [reply CHANGELOG.md]: Fixed.

File: docs/branch_and_deploy_guide.md (hunk 20 above)
Comment (mary-cleaton) [reply docs/branch_and_deploy_guide.md]: Done.

File: docs/branch_and_deploy_guide.md (hunk 21 above)
Comment (mary-cleaton) [reply docs/branch_and_deploy_guide.md]: Fixed.

File: docs/branch_and_deploy_guide.md (hunk 22 above)
Comment (mary-cleaton) [reply docs/branch_and_deploy_guide.md]: Fixed.

File: docs/branch_and_deploy_guide.md (hunk 23 above)
Comment (mary-cleaton) [reply docs/branch_and_deploy_guide.md]: Fixed.

File: docs/user_guide.md (hunk 24 above)
Comment (mary-cleaton) [reply docs/user_guide.md]: Fixed.

File: docs/user_guide.md (hunk 25 above)
Comment (mary-cleaton) [reply docs/user_guide.md]: Completed.

File: docs/user_guide.md (hunk 26 above)
Comment (mary-cleaton) [reply docs/user_guide.md:5]: Think I've fixed it.

File: docs/user_guide.md (hunk 27 above)
Comment (mary-cleaton) [reply docs/user_guide.md:28]: I think I've fixed this.

File: docs/user_guide.md (hunk 28 above)
Comment (mary-cleaton) [reply docs/user_guide.md]: Generally one shouldn't do relative links in `.md` files as they break if the `.md` file is moved. One can use `/` for "head of repo", though, which is what I've changed these to use. I think this should work. I've also updated the relative links in the other files in `docs` to also use this convention.

File: docs/user_guide.md (hunk 29 above)
Comment (mary-cleaton) [reply docs/user_guide.md]: See above.

File: .github/workflows/deploy_to_pypi.yaml (hunk 4 above)
Comment (mary-cleaton) [reply .github/workflows/deploy_to_pypi.yaml]: Oh, duh. I knew I had an example somewhere, I'd just forgotten where. Yes, that should work!
Added.

File: docs/user_guide.md (hunk 28 above)
Comment (danasebo) [reply docs/user_guide.md]: perfect solution, it works now, thanks Mary!

File: .github/workflows/deploy_to_pypi.yaml (hunk 4 above)
Comment (danasebo) [reply .github/workflows/deploy_to_pypi.yaml]: I _think_ run can be used only inside step, not outside.. not sure if this will work as is now. I would maybe suggest not to complicate it and remove it, apologies for a confusion!

**Review by danasebo — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy_to_pypi.yaml (hunk 4 above)
Comment (mary-cleaton) [reply .github/workflows/deploy_to_pypi.yaml]: Have moved it into the first step of `release-build`, so I think it should now work.

**Review by danasebo — APPROVED** [review]
Many thanks for all the updates, Mary! It looks good to me now, and I'm happy for you to merge.
